NIP-96 - HTTP File Storage Integration (#547)
nostr-protocol/nips: 96.md

NIP-96
======

HTTP File Storage Integration
-----------------------------

`draft` `optional` `author:arthurfranca` `author:Semisol` `author:staab`

## Introduction

This NIP defines a REST API for HTTP file storage servers intended to be used in conjunction with the nostr network.
The API will enable nostr users to upload files and later reference them by url on nostr notes.

The spec DOES NOT use regular nostr events through websockets for
storing, requesting nor retrieving data because, for simplicity, the server
will not have to learn anything about nostr relays.

## Server Adaptation

File storage servers wishing to be accessible by nostr users should opt-in by making available

> an https route at `/.well-known/nostr.json` with `nip96.routes` field in the following form:

Author: Ok I will change it.

```js
{
  "nip96": {
    // required
    "routes": {
      "upload": "https://upload.domain/upload-path",
      "download": "https://download.domain/download-path",
      "deletion": "https://deletion.domain/deletion-path"

>     },

Author: I though this too at first but then at the time I thought it would be easier to understand with 3 separate keys.

So you want A) or B)?

A)
```
  "routes": {
      "api": "https://api.domain/upload-and-deletion-path",
      "download": "https://download.domain/download-path"
   }
```

B)
```
  "routes": {
    "upload/deletion": "https://upload.and.deletion.domain/upload-and-deletion-path",
    "download": "https://download.domain/download-path"
  }
```

>     },

Author: Using "api_url" (that may include any custom path) and "download_url"

    // optional
    "supported_nips": [],
    // optional
    "tos_url": "https://my.landing.page/terms-of-service",
    // optional
    "plans": {
      // "free" is the only standardized plan key and
      // clients may use its presence to learn if server offers free storage
      "free": {
        "name": "Free Tier",
        "url": "https://...", // payment etc
        "max_byte_size": 10485760,
        // range in days / 0 for no expiration
        // [7, 0] means it may vary from 7 days to unlimited persistence,
        // [0, 0] means it has no expiration
        // early expiration may be due to low traffic or any other factor
        "file_expiration": [14, 90],
        "image_transformations": ["resizing"],
        "cost": {
          // it considers lowest currency unit
          // "100" USD equals 100 cents or 1 dollar
          // "1000" BTC equals 1000 milisats or 1 satoshi
          "price": "0", // string
          "currency": "USD",
          // "per-upload|one-time|day|week|month|year..."
          "frequency": ""

>         }

Author: Ok I will remove the "cost" part and leave the rest as a hint for clients to know if the server accepts free uploads (what most users expect).

      }
    }
  }

> }

Author: Yes, you version (root_url) is in fact very similar to how it was written originally. But because [of this](https://github.com/nostr-protocol/nips/pull/547#discussion_r1256083467) I had to change it.

> }

Author: People complained of reserving paths such as /file. Also PUT isn't supported by html form fields that's why I used POST.

Ended up with api_url (can include custom paths) and optional download_url.

```

Note that if using a CDN, the `server` should issue a 302 redirect from
the `nip96.routes.download` route to the CDN url instead
of assigning the CDN directly as the `nip96.routes.download` route.
This is because some NIP-96 integrations
may require NIP-98 authentication for downloads,

> which the CDN won't be able to support.

Author: Upload/download/deletion at same root_url would surely make things simpler.

The text you quoted is warning for the case of some future nip integrations (like NIP-60)
that requires NIP-98 authentication for downloads.

But for other cases such as free downloads, unfortunatelly, some hosts may want to point directly to a CDN
without the redirect to prevent files from going offline when the API is down.

Hey, @bndw @michaelhall923 and @v0l can you confirm if the api-to-cdn redirect would or not be acceptable for downloads?

> which the CDN won't be able to support.

Author: Now there is api_url (that may include any path) and optional download_url.

### List of Supporting File Storage Servers

| Name | Domain |
| ---- | ------ |

> | -    | -      |

Author: @v0l and @michaelhall923 expressed their intention to implement it.
But i guess it is a chicken and egg problem. They are probably waiting for this to get finished and merged first.

> | -    | -      |

Author: This list was meant as a kickstart list for clients (Amethyst, coracle...). With time, server hints inside the "x" tag/param and updated NIP-65? hints could make it dynamic.

The file extension may be absent because the hash is the only needed string to uniquely identify a file.

Example: `https://download.domain/download-path/719171db19525d9d08dd69cb716a18158a249b7b3b3ec4bbdec5698dca104b7b.png`

### Image Transformations

`Servers` may respond to some image transformation query parameters and ignore those they don't support by serving
the original image file without transformations.

#### Resizing

Upon upload, `servers` may create resized image variants, such as thumbnails, respecting the original aspect ratio.
`Clients` may use the `w` query parameter to request an image version with the desired pixel width.
`Servers` can then serve the variant with the closest width to the parameter value
or an image variant generated on the fly.

Example: `https://download.domain/download-path/<sha256-file-hash>.png?w=32`

## Deletion

`Servers` must make available the route `https://deletion.domain/deletion-path/<sha256-file-hash>(.ext)` (route from `https://server.example.domain/.well-known/nostr.json`) with `DELETE` method for file deletion.
The extension is optional as the file hash is the only needed file identification.

> `Clients` should send a `DELETE` request to the server url in the format `https://deletion.domain/deletion-path/<sha256-file-hash>.png`. It must include a NIP-98 `Authorization` header.

Author: Fixed

NIP-96
======

HTTP File Storage Integration
-----------------------------

`draft` `optional` `author:arthurfranca` `author:Semisol` `author:staab`

## Introduction

> This NIP defines a REST API for HTTP file storage servers intended to be used in conjunction with the nostr network.

Author: Ok I'm adding "accepted_mime_types" field to `/.well-known/nostr/nip96.json`. The user already can send a field `content_type`: "<mime type>" when uploading. The server can block what it doesn't support.

The main use case is for media storage but who knows the future.

> This NIP defines a REST API for HTTP file storage servers intended to be used in conjunction with the nostr network.

Author: Ok changing it to content_types

The API will enable nostr users to upload files and later reference them by url on nostr notes.

The spec DOES NOT use regular nostr events through websockets for
storing, requesting nor retrieving data because, for simplicity, the server
will not have to learn anything about nostr relays.

## Server Adaptation

File storage servers wishing to be accessible by nostr users should opt-in by making available an https route at `/.well-known/nostr/nip96.json` with `api_url`:

```js
{
  // Required
  // File upload and deletion are served from this url
  // Also downloads if "download_url" field is absent or empty string
  "api_url": "https://your-file-server.example/custom-api-path",
  // Optional

>   // If absent, downloads are served from the api_url

Author: "api_url" field was just to merge "delete" and "upload", depending if it is POST or DELETE action.

At this moment this NIP is just for basic functionallity (upload/delete/download one file at a time). Your server can have extra features outside of the NIP scope or we can expand it after PR merge. 

Do you need me to split delete and upload to their own urls for some reason?

>   // If absent, downloads are served from the api_url

Author: I think I got what you mean. We do need atleast an url to search for files by hash (so one can search at many servers till one finds it).

But what if we keep the download_url at `/.well-known/nostr/nip96.json` so that when trying to find a file at other servers, I can use this fixed one (or api_url when absent). **BUT** what if after uploading a file, the upload response includes an url to download the uploaded file? Now you can kind of load balance by sending different urls at each upload like:

Upload 1: response has url "https://your-file-server.example/path-A/hash1"
Upload 2: response has url "https://your-file-server.example/path-B/hash2"
Upload 3: response has url "https://your-file-server.example/path-C/hash3"
Upload 4: response has url "https://your-file-server.example/path-A/hash1"
and so on..

In the example you have a group of 3 different download routes you use so effectively load balancing, because the nostr events will use them instead of what is informed at the config file.

>   // If absent, downloads are served from the api_url

Author: Alright. You can do it here at the "url" upload response field:

https://github.com/nostr-protocol/nips/blob/581340338d93de40e8801f3eba689420b6832305/96.md?plain=1#L147-L150

It can be any hostname, any path, any filename. This url will be the one used inside nostr events to download the file.

But, for decentralization purposes, you still have to support this fixed url when the user just knows the original file hash and wanna try downloading from you:

https://github.com/nostr-protocol/nips/blob/581340338d93de40e8801f3eba689420b6832305/96.md?plain=1#L223-L225

Does this solve the problem?

>   "download_url": "https://a-cdn.example/a-path",

Author: > separation of endpoints based on a file type. If the target just a file and not media, then this may be OK.

There are libs able to detect file type just from the begginning of a file content like [https://github.com/sindresorhus/file-type](https://github.com/sindresorhus/file-type). Can't you use this to block upload of unsupported types or types that weren't able to be recognized by such a lib?

>   "download_url": "https://a-cdn.example/a-path",

Author: Yeah sorry downloads. But it is a fixed download path that the client will just append the file hash it wants to download.

There will be also temporary download urls for cases of waiting for server media processing as @quentintaranpino just suggested. But these will be inserted in the upload response instead of in this config file.

>   "download_url": "https://a-cdn.example/a-path",

Author: I will expand this topic at your conversation below.

  // Optional
  // Use this if you are a nostr relay using your /.well-known/nostr/nip96.json

>   // just to advertise someone else's http file storage server's /.well-known/nostr/nip96.json

Author: I will add a `Note: This field is not meant to be set by HTTP Servers.`

This is just if someone have the bad idea of setting up a /.well-known/nostr/nip96.json config file copying, for example, nostr.build's config file. Intead of that, it should keep an empy config file and use the field to redirect to nostr.build's config file.

I will also change the field name to "delegated_to_url"

  // In this case, "api_url" field must be an empty string
  "recommended_url": "https://your-file-server.example",
  // Optional
  "supported_nips": [],
  // Optional
  "tos_url": "https://your-file-server.example/terms-of-service",
  // Optional
  "plans": {
    // "free" is the only standardized plan key and
    // clients may use its presence to learn if server offers free storage
    "free": {
      "name": "Free Tier",
      "url": "https://...", // payment etc
      "max_byte_size": 10485760,
      // Range in days / 0 for no expiration
      // [7, 0] means it may vary from 7 days to unlimited persistence,
      // [0, 0] means it has no expiration
      // early expiration may be due to low traffic or any other factor
      "file_expiration": [14, 90],

>       "image_transformations": ["resizing"]

Author: This was just a small easy addition and it is totally optional to have it.

I think it is better to try to merge this PR before adding more features. Besides that, I have no xp on video transcoding xD.

>       "image_transformations": ["resizing"]

Author: Ok I will remove it and let it be added by you or someone else in a future PR.

    }
  }
}
```

> Note that if using a CDN, the `server` MAY issue a 302 redirect from

Author: I want to start saying that this part was just advice, totally optional.

It talks about api when the server use the simple setup of using the same API url with GET for downloads. Or else it talks about dedicated download url.

It only says that, if for example you are charging for downloads (not uploads), and want to use NIP-98 for auth, CDNs can't be used directly because they don't know about NIP-98 nor even know that nostr is a thing.

> Note that if using a CDN, the `server` MAY issue a 302 redirect from

Author: Nevermind, I saw you comment below and got that CDNs can understand the NIP-98 custom Authorization header. I will remove this paragraph.

GET requests to the `api_url` or `download_url` route to the CDN url instead
of assigning the CDN directly as the `download_url` route.
This is because some NIP-96 integrations
MAY require NIP-98 authentication for downloads,

> which the CDN won't be able to support.

Author: Oh I didn't know about that. I really thought that CDNs couldn't be configured to parse the NIP-98 custom Authorization header. I will remove this paragraph.

> ### Relay Hints

Author: I changed a little bit the paragraph and added `Note: This section is not meant to be used by HTTP Servers.`

Like I said above, "This is just if someone have the bad idea of setting up a /.well-known/nostr/nip96.json config file copying, for example, nostr.build's config file. Intead of that, it should keep an empy config file and use the field to redirect to nostr.build's config file."

A nostr relay MAY advertise someone else's HTTP file storage server by
adding a `/.well-known/nostr/nip96.json` with "recommended_url" field
pointing to the url where the server hosts its own
`/.well-known/nostr/nip96.json`. In this case, the "api_url" must
be an empty string.

If the nostr relay is also an HTTP file storage server,
it must use the "api_url" field instead.

### List of Supporting File Storage Servers

| Name | Domain |
| ---- | ------ |
| -    | -      |

## Upload

> A file can be uploaded one at a time to `https://your-file-server.example/custom-api-path` (route from `https://your-file-server.example/.well-known/nostr/nip96.json` "api_url" field) as `multipart/form-data` content type using `POST` method with the file object set to the `file` form data field.

Author: This is just bare bones minimum API to upload one file at a time using `multipart/form-data` with a POST that can be made with any HTML form.

If you need extra features, I ask you to wait for the PR merge with just this min set of features. Then we can patch the NIP with more features.

Unless you think the current setup will somehow make it not viable to add future extensions to the NIP.

`Clients` must add a [NIP-98](98.md) `Authorization` header with the encoded `payload` tag set to the base64-encoded 256-bit SHA-256 hash of the file (not the hash of the whole request body).

> If using an html form, use an `Authorization` form data field instead.

Author: If using an HTML `<form>` to send file without javascript, the client won't be able to set request headers. That's why the NIP-98 `Authorization` content will be a regular field instead of a header.

These following **optional** form data fields MAY be used by `servers` and SHOULD be sent by `clients`:
- `expiration`: string of the UNIX timestamp in seconds. Empty string if file should be stored forever. The server isn't required to honor this;
- `size`: string of the file byte size. This is just a value the server can use to reject early if the file size exceeds the server limits;

> - `alt`: strict description text for visibility-impaired users;

Author: The user may want to just send the file and then later at a NIP-94 event set an `alt` tag?

> - `alt`: strict description text for visibility-impaired users;

Author: I will atleast add (recommended) next to it. Although a shame, its hard to imagine everyone sending a meme will describe it with detail before sending it. By making it required, people may just write "HJKHSKHAHhahshhabd".

- `caption`: loose description;
- `content_type`: mime type such as "image/jpeg".

> Others custom form data fields may be used depending on specific `server` support.

Author: Anyone can send a new PR to extend this NIP or make another NIP as an extension to this one.

The `server` isn't required to store any metadata sent by `clients`.

> Note for `clients`: if using an HTML form, it is important for the `file` form field to be the **last** one, or be re-ordered right before sending or be appended as the last field of XHR2's FormData object.

Author: With `multipart/form-data` if the file comes before the fields, you won't be able to fail early based on the fields content. The server would have to download the whole file first before receiving the fields.

> Note for `clients`: if using an HTML form, it is important for the `file` form field to be the **last** one, or be re-ordered right before sending or be appended as the last field of XHR2's FormData object.

Author: The length can't be trusted but if the client says, using the `size` field, that the file has 10MB while the server limit is 5MB, the server has the option to reject it early before stating the file download. It depends on lib support to analyze fields before receiving the file.

> The `filename` embedded in the file may not be honored by the `server`, which could internally store just the SHA-256 hash value as the file name, ignoring extra metadata.

Author: This will be possible with the x tag appended to the url (with NIP-54) or with the x tag (with NIP-94).
We can't use the filename because other file urls not using NIP-96 most times won't use the convention of having the SHA-256 as filename.

> The `filename` embedded in the file may not be honored by the `server`, which could internally store just the SHA-256 hash value as the file name, ignoring extra metadata.

Author: > SHA-256 of the file content (before any alterations) seems to be the good starting point.

In practice that's what this PR wants. A client must be able to download at `https://download.route/path/<original-file-hash>`. In the future when the file gets deleted, a client can try another server's download route appending the original file's hash as if it was the filename.

The hash is enough to uniquely identify a file, that's why it will be used on the "download" and "delete" routes.

> The `server` MUST link the user's `pubkey` string (which is embedded in the decoded header value) as the owner of the file so to later allow him to delete the file.

Author: ok will edit it

Note that if a file with the same hash of a previously received file (so the same file) is uploaded by another user, the server doesn't need to store the new file.

> It should just add the new user's `pubkey` to the list of the owners of the already stored file with said hash.

Author: When the server stores the file a user X sent and link it to a SHA-256 hash, another user Y may upload the same file that results in the same SHA-256 hash.

Your server do not need to store the same file twice. The server just have to know  user X owns it and user Y owns it too. When Y deletes it, now just user X owns it (but the file isn't deleted). When user X deletes it, now no one owns it, so the file gets deleted.

However your server is free to store two copies of the same file (one foe each user that upload it) if its is better for you.

This answers your question?

> The `server` SHOULD also store the `Authorization` header/field value (decoded or not) for accountability purpose as this proves that the user with the unique pubkey did ask for the upload.

Author: This is just a "SHOULD", so it is optional. It means that the NIP-98 event is a signed proof of everything that is inside the event. So if there is a SHA-256 of the file inside the event, for example, it proves without doubt that the npub sent it.

> The `server` SHOULD also store the `Authorization` header/field value (decoded or not) for accountability purpose as this proves that the user with the unique pubkey did ask for the upload.

Author: Changed to "The `server` MAY also store the `Authorization` header/field value (decoded or not) for accountability purpose as this proves that the user with the unique pubkey did ask for the upload of the file with a specific hash. However, storing the pubkey is sufficient to establish ownership."

The `server` MUST reject with 413 Payload Too Large if file size exceeds limits.

The `server` MUST reject with 400 Bad Request status if some fields are invalid.

> The `server` MUST reject the upload with 403 Forbidden status if the `payload` tag value contains an already used SHA-256 hash (if file is already owned by the same pubkey) or it isn't the same of the received file.

Author: Good idea

The `server` MAY reject the upload with 402 Payment Required status if the user has a pending payment (Payment flow is not strictly required. Server owners decide if the storage is free or not. Monetization schemes may be added later to correlated NIPs.).

The upload response is a json object as follows:

```js
{

>   nip96: {

Author: ok changing it

```js
[
  // Note we appended .png from the original uploaded file extension to the received `x` value
  // (it is optional but extremely recommended to add the extension as it will help nostr clients with detecting the file type by using regular expression)
  ["url", "https://your-file-server.example/custom-api-path/719171db19525d9d08dd69cb716a18158a249b7b3b3ec4bbdec5698dca104b7b.png"],
  [
    "x",
    "719171db19525d9d08dd69cb716a18158a249b7b3b3ec4bbdec5698dca104b7b",
    // Server hostname (from nip96.hint_url response)
    // where one can find the /.well-known/nostr/nip96.json config resource.
    "https://your-file-server.example"
  ]
]
```

The `x` tag's third value is an important hint that clients can use to find new NIP-96 compatible file storage servers.

Alternatively, instead of using NIP-94, the `client` can share or embed on a nostr note just the above url with added "x" [NIP-54](54.md) inline metadata field values and optionally other fields such as "m" for mime type.

### File compression

> File compression is optionally done at the `client`, **never on server**, before uploading.

Author: Changing it to "File compression and other transformations like metadata stripping can be applied by the server. However, for all file actions, such as download and upload, the **original** file SHA-256 hash is what identifies the file in the url string."

> This is done to assure the same file hash result, no matter to how many servers one may upload the same file.

Author: You are right, changed it.

## Download

> `Servers` must make available the route `https://your-file-server.example/custom-api-path/<sha256-file-hash>(.ext)` (route from `https://your-file-server.example/.well-known/nostr/nip96.json` "api_url" or "download_url" field) with `GET` method for file download.

Author: Sorry this isn't a redirect. It is just the url value of the "download_url" field inside the `https://your-file-server.example/.well-known/nostr/nip96.json` config file.

The config file allows a server to choose any custom path even at another domain for the download_url.

Supporting ".ext", meaning "file extension", is required for `servers`. It is optional, although recommended, for `clients` to append it to the path.
When present it may be used by `servers` to know which `Content-Type` header to send (e.g.: "Content-Type": "image/png" for ".png" extension).
The file extension may be absent because the hash is the only needed string to uniquely identify a file.

Example: `https://your-file-server.example/custom-api-path/719171db19525d9d08dd69cb716a18158a249b7b3b3ec4bbdec5698dca104b7b.png`

### Image Transformations

`Servers` may respond to some image transformation query parameters and ignore those they don't support by serving
the original image file without transformations.

#### Resizing

Upon upload, `servers` may create resized image variants, such as thumbnails, respecting the original aspect ratio.
`Clients` may use the `w` query parameter to request an image version with the desired pixel width.
`Servers` can then serve the variant with the closest width to the parameter value
or an image variant generated on the fly.

> 

Author: > Transforming to any arbitrary dimensions should not be encouraged

Dynamic transformations is common in services like [https://cloudinary.com/documentation/image_transformations](https://cloudinary.com/documentation/image_transformations).

Though the text says that the server can generate thumbnails at any dimensions and "serve the variant with the closest width to the parameter value". So the server doesn't need to support dynamic transformations as it probably should be a paid service.

> 

Author: Oh I had removed based on your comment elsewhere. Re-adding it then...

storing, requesting nor retrieving data because, for simplicity, the server
will not have to learn anything about nostr relays.

## Server Adaptation

File storage servers wishing to be accessible by nostr users should opt-in by making available an https route at `/.well-known/nostr/nip96.json` with `api_url`:

```js
{
  // Required
  // File upload and deletion are served from this url
  // Also downloads if "download_url" field is absent or empty string
  "api_url": "https://your-file-server.example/custom-api-path",
  // Optional
  // If absent, downloads are served from the api_url
  "download_url": "https://a-cdn.example/a-path",
  // Optional
  // Note: This field is not meant to be set by HTTP Servers.
  // Use this if you are a nostr relay using your /.well-known/nostr/nip96.json
  // just to redirect to someone else's http file storage server's /.well-known/nostr/nip96.json
  // In this case, "api_url" field must be an empty string
  "delegated_to_url": "https://your-file-server.example",
  // Optional
  "supported_nips": [60],
  // Optional
  "tos_url": "https://your-file-server.example/terms-of-service",
  // Optional

>   "content_types": ["image/jpeg", "video/webm"]

Reviewer: From my point of view it is correct.

>   "content_types": ["image/jpeg", "video/webm"]

Author: I will add image/* to the example

NIP-96
======

HTTP File Storage Integration
-----------------------------

`draft` `optional` `author:arthurfranca` `author:Semisol` `author:staab`

## Motivation

This NIP defines a simple unified REST API for HTTP file storage servers so to integrate them to the nostr network.
The API will enable nostr users to upload files and later reference them by url on nostr notes.

The spec DOES NOT use regular nostr events through websockets for
storing, requesting nor retrieving data because, for simplicity, the server
will not have to learn anything about nostr relays.

## Server Adaptation

File storage servers wishing to be accessible by nostr users should opt-in by making available
an https route at `/nip96(/<sha256-file-hash>)` for handling files as described below.

### List of Supporting File Storage Servers

| Name | Domain | Upload Cost | NIP-60 Fee | File Retention Policy |
| ---- | ------ | ----------- | ---------- | --------------------- |
| -    | -      | -           | -          | ❌️                    |

## Upload

> A file can be uploaded one at a time to `https://server.domain/nip96` as `multipart/form-data` content type using `POST` method with the file object set to the `file` form data field.

Reviewer 2: ```suggestion
A file can be uploaded one at a time to `https://server.domain/.well-known/nip96` as `multipart/form-data` content type using `POST` method with the file object set to the `file` form data field.
```

It's a big assumption that servers can simply add a new route. For example, Twitter would have to delete this user's account: https://twitter.com/nip96

There are all kinds of problems with exposing paths on the top-level. People will have to update their Service Workers and caches. At least `.well-known` is well-known, and usually already has some special handling.

Curious what @melvincarvalho thinks about this.

Reviewer 2: Also, I think this NIP can evolve by having relays add nip96 to their supported_nips and expose the nip96 endpoint. Then clients can select a relay with nip96 support to upload files to.

Reviewer 3: isn't .well-known more used for metadata stuff? not really where you'd put an upload route.

what about putting a nip96 object in /.well-known/nostr.json that contains info about the routes?

Author: Using  /.well-known/nostr.json as config now

It should just add the new user's `pubkey` to the list of the owners of the already stored file with said hash.

The `server` SHOULD also store the `Authorization` header/field value (decoded or not) for accountability purpose as this proves that the user with the unique pubkey did ask for the upload.

The `server` MUST reject with 413 Payload Too Large if file size exceeds limits.

The `server` MUST reject with 401 Unauthorized status if some fiels are invalid.

The `server` MUST reject the upload with 403 Forbidden status if the `payload` tag value contains an already used SHA-256 hash (if file is already owned by the same pubkey) or it isn't the same of the received file.

The `server` MAY reject the upload with 402 Payment Required status if the user has a pending payment (Although payment flow is not strictly required. Server owners decide if the storage is free or not. Although we suggest [this freemium monetization scheme](#monetization))

The upload response is a json as follows:

```js
{
  nip96: {
    // SHA-256 hash. Empty string if unsuccessful

>     x: "719171db19525d9d08dd69cb716a18158a249b7b3b3ec4bbdec5698dca104b7b",

Reviewer 4: It would be nice if the server could return a canonical URL for the hosted media. This would allow e.g. uploads to `api.example.com`, while serving the media from `cdn.example.com`. But the only way that works is if the server can tell the client where the media is accessed.

Author: The download is always at `api.example.com/nip96/the-x-value` both with or without a file extension

Reviewer 4: Right. But imagine I have a CDN that I want to serve the files from that is separate from the API that accepts the uploads. If the server could tell the client the canonical URL for the media, you could facilitate this.

Reviewer 3: yeah you should be able to have different hostnames

Author: @bndw @michaelhall923 Oh, sorry misunderstood. Now I get it.

However, what happens if the server using `cdn.example.com/...` for downloads in the future starts using another cdn and changes download urls to `anothercdn.example.com/...`? This wouldn't be a problem if the urls were used on a website you control but it does is a problem for urls used on nostr notes.

That's why in my opinion the best approach is for the server to simply issue a 302 redirect from the fixed download url to whatever cdn it wants.

Or what you want is a different hostname for the fixed url that then would issue the 302 redirects to the cdn? Like the upload (at `a-server.com/nip96`) response would have something like this:
```js
{
  nip96: {
    hostname: 'another.hostname.com' // not a cdn
    x: "719171db19525d9d08dd69cb716a18158a249b7b3b3ec4bbdec5698dca104b7b",
  },
  errors: {  nip96: [] }
}
```

Then you would be able to do a 302 redirect from `another.hostname.com/nip96/719171...` (instead of from `a-server.com/nip96/719171...`) to `cdn.example.com/nip96/719171...`?

The downside is some servers may wrongly reply with `{ nip96: {  hostname: 'cdn.example.com' } ... }` pointing directly to the cdn.

Reviewer 4: Yeah the core of the idea is that GET requests would never hit the upload server. This has nice properties around availability and caching e.g. if the upload server (api.example.com) gets overwhelmed it does not have to impact serving files (cdn.example.com).

Author: This is now configurable at  /.well-known/nostr.json.
The response also has a "download_route" field (can add custom paths to, so no /nip96 path anymore)

the original image file without transformations.

#### Resizing

Upon upload, `servers` may create resized image variants, such as thumbnails, respecting the original aspect ratio.
`Clients` may use the `w` query parameter to request an image version with the desired pixel width.
`Servers` can then serve the variant with the closest width to the parameter value
or an image variant generated on the fly.

Example: `https://download.domain/download-path/<sha256-file-hash>.png?w=32`

## Deletion

`Servers` must make available the route `https://deletion.domain/deletion-path/<sha256-file-hash>(.ext)` (route from `https://server.example.domain/.well-known/nostr.json`) with `DELETE` method for file deletion.
The extension is optional as the file hash is the only needed file identification.

`Clients` should send a `DELETE` request to the server url in the format `https://deletion.domain/deletion-path/<sha256-file-hash>.png`. It must include a NIP-98 `Authorization` header.

The `server` should reject deletes from users other than the original uploader. The `pubkey` encoded on the header value identifies the user.

> It should be noted that more than one user may have uploaded the same file (so with the same hash). In this case, a delete must not really delete the file but just remove the user's `pubkey` from the file owners list.

Reviewer 5: ```suggestion
It should be noted that more than one user may have uploaded the same file (with the same hash). In this case, a delete must not really delete the file but just remove the user's `pubkey` from the file owners list.
```

The response is similar to the json sent on uploads, but without the `nip96.download_route` field.

## Monetization

This is an **optional** step that may be configured by `servers` at any moment.
It is a suggested way of monetizing the file storage.

It is a freemium business model, meaning that the file storage is free,
but it is expected that some users consuming the stored files may "zap" (send)
a variable amount of cryptocurrency to the server owner wallet.

Although file storage is free of charge, of course the `server` can delete files at any moment at its discrition.

> For example, the file storage `server` may fetch [zap receipts](57.md#appendix-e-zap-receipt-event) from it's zap provider pubkey and
> extract embeded media from referenced notes. Then stored files that hasn't received zaps
> for a while may be deleted.

Reviewer 5: ```suggestion
For example, the file storage `server` may fetch [zap receipts](57.md#appendix-e-zap-receipt-event) from its
zap provider pubkey and extract embedded media from referenced notes. Stored files that haven't received zaps
for a while may be deleted.
```

### Server Configuration

`Server` owners should configure their nostr pubkey at the `/.well-known/nostr.json?name=_` path as per [NIP-05](05.md#showing-just-the-domain-as-an-identifier), including `relay` information.

`Server` owners then should add to their nostr account metadata event (kind 0) a lud06 or lud16 field (and/or any future supported cryptocurrency address) pointing to a `zap-enabled` wallet address they own.

> It is required to send the event to all relays included on previous NIP-05 configuration. We suggest re-sending the same event to said relays from time to time to make sure they don't get deleted.

Reviewer 5: This seems iffy to me, maybe adding support for lightning addresses on the nostr.json endpoint would be preferable?

Author: Totally agree! Http file storage servers shouldn't need websockets, just `/.well-known/nostr.json` config. But this [PR](https://github.com/nostr-protocol/nips/pull/552/files) tells that the zap tag can only have pubkey as value. It needs a corresponding kind 0 event to get the updated lud06/16 address.

If @vitorpamplona agrees to update the PR to make the zap tag also accept a NIP-05 address, the server could use just the `https://my.file.server/.well-known/nostr.json` to config the current lud06/16 address.

Like this:
```js
    ["zap", "82341f882b6eabcd2ba7f1ef90aad961cf074af15b9ef44a09f9d2a8fbfbe6a2", "wss://nos.lol", "19"],
    ["zap", "_@my.file.server", "", "1"]
```

Author: @vitorpamplona L@_@K

Author: It was tried and rejected [here](https://github.com/nostr-protocol/nips/pull/696)

  ],
  "content": "<description>",
  "sig": "<signature>" // signed with private key linked to the above pubkey
}
```

### Download

The `server` serves downloads as usual but it additionally requires authentication using NIP-98.
The event encoded on `Authorization` header should be similar to the following:

```js
{
  "id": "<32-bytes lowercase hex-encoded sha256 of the the serialized event data>",
  "pubkey": "<32-bytes lowercase hex-encoded public key of the user>",
  "created_at": "<unix timestamp in seconds>",
  "kind": 27235,
  "tags": [
    [
        "u",
        "https://download.domain/download-path/719171db19525d9d08dd69cb716a18158a249b7b3b3ec4bbdec5698dca104b7b.json" // with or without the extension
    ],
    [
        "method",
        "GET"
    ]
  ],
  "content": "",
  "sig": "<64-bytes hex of the signature of the sha256 hash of the serialized event data, which is the same as the id field>"
}
```

The `server` should check if the pubkey has already payed for the content, i.e., the sum of the related NIP-57 zap receipts
must be greater than or equal to the price set on upload and the `server` must have received its share.

> Its advisable to store payer pubkeys so to fast check next time if the user is authorized to download the file.

Reviewer 5: ```suggestion
It's advisable to cache payer pubkeys for faster authorization the next time the user attempts to download the file.
```

The `filename` embedded in the file may not be honored by the `server`, which could internally store just the SHA-256 hash value as the file name, ignoring extra metadata.
The hash is enough to uniquely identify a file, that's why it will be used on the "download" and "delete" routes.

The `server` MUST link the user's `pubkey` string (which is embedded in the decoded header value) as the owner of the file so to later allow him to delete the file.
Note that if a file with the same hash of a previously received file (so the same file) is uploaded by another user, the server doesn't need to store the new file.
It should just add the new user's `pubkey` to the list of the owners of the already stored file with said hash.

The `server` SHOULD also store the `Authorization` header/field value (decoded or not) for accountability purpose as this proves that the user with the unique pubkey did ask for the upload.

The `server` MUST reject with 413 Payload Too Large if file size exceeds limits.

> The `server` MUST reject with 401 Unauthorized status if some fiels are invalid.

Reviewer 4: nit: This should probably be `400 Bad Request`

Author: Thanks. Fixed typo and changed to 422.

Reviewer 4: FWIW I have never encountered a 422 in the wild. Pretty much everyone uses a 400 for Bad Requests.

Author: Ok changed 422 to 400
